master/internal/api_project.go

```diff
@@ -345,3 +345,57 @@ func (a *apiServer) UnarchiveProject(
 	}
 	return &apiv1.UnarchiveProjectResponse{}, nil
 }
+
+func (a *apiServer) GetProjectsByUserActivity(
+	ctx context.Context, req *apiv1.GetProjectsByUserActivityRequest,
+) (*apiv1.GetProjectsByUserActivityResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	p := []*model.Project{}
+
+	err = db.Bun().NewSelect().Model(p).NewRaw(`
+	WITH p as (
+		SELECT  pr.*, activity_time FROM projects pr 
+		JOIN activity a ON pr.id = a.entity_id
+	),
+	pe AS (
+	  SELECT project_id, state, start_time
+	  FROM experiments
+	)
+	SELECT w.name as workspace_name, u.username, p.id, p.name, 
+	p.workspace_id, p.description, p.immutable, p.notes, p.user_id,
+			'WORKSPACE_STATE_' || p.state AS state, p.error_message,
+	  SUM(case when pe.project_id = p.id then 1 else 0 end) AS num_experiments,
+	  SUM(case when pe.project_id = p.id 
+		AND pe.state = 'ACTIVE' then 1 else 0 end) AS num_active_experiments,
+	  MAX(case when pe.project_id = p.id then pe.start_time else NULL end) 
+	  AS last_experiment_started_at
+	FROM pe, p
+	  LEFT JOIN users as u ON u.id = p.user_id
+	  LEFT JOIN workspaces AS w on w.id = p.workspace_id
+	GROUP BY p.user_id, p.id, p.name, p.workspace_id, p.description, 
+	p.immutable, p.notes, p.state, p.error_message,  u.username, w.name, p.activity_time
+	ORDER BY p.activity_time DESC;`).
```

Comment:
how do we control the size of the `activity` table? say I am using the product for an extended period of time, touching hundreds of projects over a couple years. 
- should we cleanup old activity?
- should we limit the number of returned projects here to some number, or have pagination for this API?

Reply:
cc: @gt2345 since she will be taking over from here. 

I do think we should support pagination in some form we will only ever be showing a fixed number (I am not sure if we will be allowing users to change this number manually). I think we might only even need a limit and not necessarily an offset since we might just be showing a fixed size list.

I'd have to recheck with @alexjdetermined but I think we always want a user to see their "most recent" activity whether that was 1 minute or 1 year ago. So we might not be able to delete "old" entries. But instead we'd need some logic such as "keep at least X number of the most recent entries" so the dashboard can always be populated fully. 

I do think pagination should be implemented in some form in the PR even if it is a fixed size limit.

Reply:
I see. if there's no offset support then probably real pagination is not necessary.